### PR TITLE
Optimize wieldmesh

### LINKED ISSUE #3069: Many wieldmesh polygons
The generated wielditem mesh has lots of unneeded polygons, which are usually not visible:
pic
pic
Would disabling adding them (searching the corners of the texture) cause more or less lag?
https://github.com/minetest/minetest/blob/master/src/wieldmesh.cpp#L69

### LINKED ISSUE #6720: Dropped items have a lot of faces
I accidentally stepped into a dropped egg and was surprised that there are so many faces inside. Is this on purpose? Lots of dropped items may cause performance issues if all these faces have to be drawn

screenshot_20171130_155806

### PR DESCRIPTION
**Pull request by numberZero**
Fixes #3069, fixes #6720.

This PR changes handling of extruded images. Instead of using single mesh for all textures (of a single size), separate optimized mesh is used for each texture. Optimization is simple and fast, working in *O(texture_pixel_count)* time, and resulting mesh uses *O(sqrt(texture_pixel_count))* memory. `createExtrusionMesh` caches and returns that mesh, but actual work is done by the `Extruder` class.

As the whole extrusion code is new, it is put in separate files; existing ones are updated minimally to support or to use it.

Note: soft overlays are not extruded at all in this PR. This may cause problems with e.g. colored books *(from some mod that I don’t remember),* but extrusion would cause Z-fighting unless done in a special, more complicated way.
Object with an overlay

### CODE REVIEW

**Comment by numberZero** [comment]
@MarkuBu Can you test this? (Note that overlays are not supported yet).

**Comment by MarkuBu** [comment]
Looks odd here
https://youtu.be/K3utU2H1OHk

Ubuntu Linux 17.10 with Nvidia driver, if that is important

**Comment by numberZero** [comment]
@MarkuBu Sorry, I forgot to commit one important fix. Try now.

**Comment by MarkuBu** [comment]
Made a short test with 16px and 64px texture. Looks good so far

:+1:

**Comment by SmallJoker** [comment]
They now look a lot thinner. Is that intended?
grafik

**Comment by numberZero** [comment]
@SmallJoker No, that was accidental. Thanks for pointing that out.

**Review by rubenwardy — COMMENTED** [review]
Code lgtm on a code review. Just two code style things:

1. Please use lowerCamelCase for functions
2. I thought create side and create face did the opposite things on first read. I suggest naming to createFrontBackSurface and createEdgeFace respectively

**Comment by rubenwardy** [comment]
This should be checked to make sure it doesn't negatively impact performance

**Comment by numberZero** [comment]
@rubenwardy On my PC it doesn’t seem to make any difference. Both in `master` and with this PR frame rate varies a lot.

**Comment by MarkuBu** [comment]
I think you need a lot of dropped items, e.g. if you flood a wheat field with water, to see a difference.

On mobile devices or built-in GPUs you may have a difference

**Comment by rubenwardy** [comment]
I'm not seeing any improvement with `3*15*15` entities spawned: 



Could be worth trying HD textures

**Comment by ThomasMonroe314** [comment]
You might not see much of a difference with sapling textures or something similar, but you would probably see a difference when you are working with something like eggs or paper

**Comment by rubenwardy** [comment]
An update: my computer is now completely frozen

Also, that's a good point as it's internal faces that matter, so you want to maximise unbroken surface area

**Comment by rubenwardy** [comment]
now testing with a square rect in a 512x512 image.



dtime goes from 87/88 to 71/71, so an improvement

```lua
minetest.register_chatcommand("a", {
	func = function(name, param)
		local r = tonumber(param)
		if not r or r <= 0 then
			r = 15
		end

		local d = 2
		local int = 1 / d

		minetest.chat_send_all("Spawning " .. (r * r * d) .. " entities, with r=" .. r .. " and int=" .. int .. ")")

		minetest.after(0.1, function()
			local pos = minetest.get_player_by_name(name):get_pos()
			for x = 0, 2*r, int do
				for z = 0, 2*r, int do
					minetest.add_item({ x = x + pos.x, y = pos.y, z = z + pos.z}, "default:diamond")
				end
			end

			minetest.chat_send_all("Done!")
		end)
	end
})

minetest.registered_entities["__builtin:item"].try_merge_with = function()
	return false
end
```

**Review by rubenwardy — DISMISSED** [review]

**Comment by numberZero** [comment]
Just remembered: items with animated tiles may be rendered wrongly. Dunno were they rendered properly with the old code, though.

**Comment by MarkuBu** [comment]
@rubenwardy what GPU and CPU do you have?

**Comment by SmallJoker** [comment]
After filling a gigabyte of swap with `r = 15`, `d = 2` using a 512x512px texture (yes, stupid me), I tested it again with `r = 5`, ensuring that enough RAM is free to get valid results.
Sadly, I can only report a performance drop with this PR:
screenshot_20171222_184817
Here's how it was before: (screenshot shows a build from Dec 1, bit shows the same as with fd9f195)
screenshot_20171222_184907

**Comment by rubenwardy** [comment]
That's with lots of edges, hmmm. Should have tested that more

> @rubenwardy what GPU and CPU do you have?

i5-5200U @ 2.7GHz, Intel HD Graphics 5500

**Comment by numberZero** [comment]
Stupid **me**. I haven’t checked the old code thoroughly. It haven’t been creating a cube for each and every pixel. Instead, it have been creating a quad for each row and for each column.

**Comment by MarkuBu** [comment]
I have an improvement at FPS. AMD FX 6300 8GB, Nvidia GTX750 2GB

default
screenshot_20171222_191306

numberZero
screenshot_20171222_191553

**Comment by numberZero** [comment]
Okay, maybe for some textures and some hardware that *is* more efficient. So there may be an option.

**Comment by numberZero** [comment]
Changed extrusion algorithm. Now should work better. @SmallJoker @rubenwardy can you test again, please?

**Comment by SmallJoker** [comment]
Now it's slightly faster than the old code:. With this PR:
screenshot_20171224_174604
Without:
screenshot_20171224_174712

**Comment by MarkuBu** [comment]
I have 20FPS vs. 32FPS with a 32px texture and 25FPS vs. 32FPS with the default 16px diamond.

I tried it with a more complex texture, a 32px sapling. The improvement is lower. 20FPS vs. 23FPS. But still better than the original. There are a lot of faces, but less faces than with default

**Comment by numberZero** [comment]
Does any item use overlay texture? In this PR, overlays have special handling; that may or may not be desired.

**Comment by ThomasMonroe314** [comment]
hmm there might be some obscure mods that have them, but I haven't noticed any for most mods.

**Comment by ExeterDad** [comment]
The mobs_redo spawn egg is made from overlays. Also the superpick is made from overlays. Is that what you're asking?

**Comment by numberZero** [comment]
@ExeterDad No, I meant hardware overlays. They can’t be used with craftitems, though; probably only `plantlike` can ever have extruded image with overlays.
screenshot_20180102_222932

**Comment by juhdanad** [comment]
@numberZero colored book covers are a possible use case (like in #6107). But I don't know if currently somebody is using this feature. Sofar's luscious mod is an example for colored overlays, because the grass node can be dropped as an item.

**Comment by juhdanad** [comment]
Also: for overlays to work properly, you have to render them with exactly the same vertex coordinates as the base's coordinates. You could require overlays to not cover areas where the base texture is transparent. That way you could just render the extruded mesh with the overlay texture.

**Comment by ThomasMonroe314** [comment]
@juhdanad if overlays do exceed the bounds of the base image, then that is the creator's problem, not the engine's

**Comment by numberZero** [comment]
@juhdanad Oh, didn’t know that overlays are supported for craftitems now. But these seem to be supported well by this patch:
screenshot_20180103_181317
Red thing is `plantlike`, so overlay is colored, but green one is a craftitem, so its overlay is not colored.

File: util/travis/clang-format-whitelist.txt
Comment (juhdanad) [inline util/travis/clang-format-whitelist.txt]: I think new files should be well-formatted, why do you exclude them?

You can use
```c++
// clang-format off
// clang-format on
```
around lines which have a special format. (based on [this](https://clang.llvm.org/docs/ClangFormatStyleOptions.html) website)

Comment (numberZero) [reply util/travis/clang-format-whitelist.txt]: > I think new files should be well-formatted, why do you exclude them?

Because linter is stupid.

> You can use
>
>     // clang-format off
>     // clang-format on

Thanks. That helps.

**Comment by numberZero** [comment]
Travis fails again.

**Comment by paramat** [comment]
Nevermind Travis, if it's broken we don't wait for it.

**Comment by juhdanad** [comment]
This implementation is faster for me too, 19-20 FPS vs. 14-16 FPS. (using 450 diamond instances)

**Comment by paramat** [comment]
Sorry for delay in reviews, much appreciated optimisation.

File: src/client/meshgen/extrusion.cpp
Comment (juhdanad) [inline src/client/meshgen/extrusion.cpp]: Is it a good idea to store this as a global variable? In my opinion creating a cache class and adding it to Client as a member (like `TextureSource`, `ShaderSource`) is a better idea (yes, I know that it was global before too).

**Review by juhdanad — COMMENTED** [review]
The code looks good overall. I just really miss documentation. But great job anyways!

File: src/client/CMakeLists.txt
Comment (SmallJoker) [reply src/client/CMakeLists.txt]: Same for `build/android/jni/Android.mk`

File: src/client/meshgen/extruder.h
Comment (SmallJoker) [inline src/client/meshgen/extruder.h]: Please initialize `data` (and optionally `texture`) to `nullptr` for safety reasons.

**Comment by paramat** [comment]
@numberZero i feel bad about how slow PR reviews are going at the moment, and you have so many good ones, sorry about this, devs seem to have less time for MT at the moment.

I'm a little concerned to see you involved with 'Minecore', i'm concerned that you may be feeling under-appreciated by MT or frustrated by slow dev. I want to state that your work is good and very much appreciated, i consider you top choice for a new MT core dev and have asked other devs to consider you.

File: src/client/meshgen/extruder.h
Comment (numberZero) [reply src/client/meshgen/extruder.h]: what safety? `texture` is set in the ctor’s initializers list, it just may not be not set. `data` is set a bit later, but still in the ctor. I updated it to throw instead of returning in the case of error, so an object without `data` set just can’t exist.

**Comment by rubenwardy** [comment]
I'm positive about that PR, however I feel that it needs lots of testing across different levels of computer

@juhdanad @Markubu @SmallJoker what are your specs?

**Comment by NathanSalapat** [comment]
Using the torch from minimal dev I tried a build of this and the latest dev and saw no difference in fps. I'll probably try again with a much higher resolution image.

System Specs:
Linux Mint 18.3 (64 bit)
Intel i5 @ 3.2 GHz
32 GB Ram
Nvidia 1060 GPU.

**Comment by NathanSalapat** [comment]
I ran another test, using the image rubenwardy provided, though I scaled it up to 128px and tried it in this build and the dev. FPS was the same but I did notice some other differences.

This build:
RTT: 0.015
mainloop_dtime 0.147 - 0.081
mainloop_draw 0.139 - 0.075

Dev build:
RTT: 0.001
mainloop_dtime 0.15 - 0.082
mainloop_draw 0.144 - 0.076

To be clear I actually don't kow what the RTT or mainloop stuff means, but I thought I'd include it for the people that do know what it means.

**Comment by SmallJoker** [comment]
@rubenwardy My specs are [here](https://github.com/minetest/minetest/pull/6803#issuecomment-353647440) (this page). Overall, it's slightly faster now on my machine - but the difference isn't too large.  It's also good to move the code into `src/client/`.

File: src/client/CMakeLists.txt
Comment (numberZero) [reply src/client/CMakeLists.txt]: done

File: src/client/meshgen/extruder.cpp
Comment (numberZero) [reply src/client/meshgen/extruder.cpp]: done

File: src/client/meshgen/extrusion.cpp
Comment (numberZero) [reply src/client/meshgen/extrusion.cpp]: The only point for that may be readability, as it won’t work any better than the current code. But having to pass `client` here and there all the time is too annoying IMO. As well as using interfaces that have only one implementation, and are very unlikely to get another one.

File: src/client/meshgen/extruder.h
Comment (numberZero) [reply src/client/meshgen/extruder.h]: done

File: src/client/meshgen/extruder.h
Comment (numberZero) [reply src/client/meshgen/extruder.h]: done, in `extrude()` description, where it is more appropriate IMO.

**Comment by juhdanad** [comment]
Much better. I would officially approve this if I could. :+1:
By the way, I really like your commit titles.

**Comment by juhdanad** [comment]
My system specs:
Ubuntu 16.04 LTS (64 bit)
Intel® Core™ i3-3120M CPU @ 2.50GHz × 4 
Intel® Ivybridge Mobile
3,7 GiB RAM

**Review by SmallJoker — DISMISSED** [review]

**Comment by numberZero** [comment]
Is there any interest in this? Or maybe refactoring without much changes is more suitable?

**Comment by paramat** [comment]
I'm interested, it's just the usual issue of finding time and energy for reviewing. It's much appreciated and much needed. Sorry dev is so quiet recently. I'll review if i can.
It might be best to not rebase until there is another near-approval, to avoid repeated rebasing for no reason.

**Comment by paramat** [comment]
Please could you add a description summary to help reviewers?

File: src/mesh.cpp
Comment (numberZero) [inline src/mesh.cpp]: This is unrelated bugfix. Can you merge it (48a2dadc981aa142781d7eb30bfa352ac89c1930) separately?

Comment (SmallJoker) [reply src/mesh.cpp]: That's possible yes. Cherry-pick or applying the diff/patch works.
Which issue does it fix?

**Comment by paramat** [comment]
Thanks, will test and will probably +1 since juhdanad's review counts for something too.

File: src/mesh.cpp
Comment (numberZero) [reply src/mesh.cpp]: Currently, calling `cloneMeshBuffer` on a buffer using `EVT_2TCOORDS` vertex type causes reads beyond its storage, thus leading to wrong data read or SEGV. That never happened only because MT doesn’t use such buffers at all, but they may be used in the future (e.g. #6820 uses them IIRC, so I had to fix that there).

**Comment by numberZero** [comment]
After the story with SAOM, I feel it’s really better to do things the right way: refactor first, optimize second. I’ll work on that, making #6874 independent of this (there are several possible ways to optimize the mesh, after all) and maybe reducing its scope. Will also make the optimization optional in this PR, for the case of RAM-limited systems (as the constant in the *O* is rather large apparently) at least.

**Comment by paramat** [comment]
So i assume WIP and don't review this yet.

**Comment by paramat** [comment]
@numberZero if it's best to refactor first, and considering the amount of conflicts this PR has, surely it would be better to close this PR and redo it after the refactor?

**Comment by numberZero** [comment]
@paramat Sure. I don’t plan to work on this PR. I don’t even remember how does it work.

Actually I don’t want to open new PR either, unless that’s what’s wanted *currently.* I can write the code, or discuss something specific, but I don’t plan scanning through GH and IRC all the time to find out the current development direction, or lack thereof (is that still the case, btw?).

**Comment by SmallJoker** [comment]
@numberZero Even though I could not notice a large improvement in terms of speed, you did a good job in cleaning up `wieldmesh.cpp`. All extrusion-related functions would now be combined, instead of scattered in the wieldmesh code.
>  I feel it’s really better to do things the right way: refactor first

Seems to be fine right now. What else would you like to refactor? Is it that it lacks soft overlay support?
> Note: soft overlays are not extruded at all in this PR.

The PR worked well on my side, but wasn't sure when and what you want to change. Using more RAM is acceptable for speed improvements; I don't expect thousands of extruded meshes to be rendered at once in regular gameplay.

**Comment by kilbith** [comment]
Worth reopening. @hecktest @jordan4ibanez

**Comment by jordan4ibanez** [comment]
I think it's a good benchmark for low end pcs when the game runs at 10 fps with quite a few item stacks on the ground, I say keep it

**Comment by oong819** [comment]
I wish this got reopen, my device lag with item mesh

**Pull request closed without merging**
